src/app/clusters/operational-credentials-server/operational-credentials-server.cpp

```diff
@@ -314,15 +315,123 @@ bool emberAfOperationalCredentialsClusterAddOpCertCallback(chip::app::Command *
                                                            chip::ByteSpan ICACertificate, chip::ByteSpan IPKValue,
                                                            chip::NodeId CaseAdminNode, uint16_t AdminVendorId)
 {
-    EmberAfStatus status = EMBER_ZCL_STATUS_FAILURE;
+    EmberAfStatus status = EMBER_ZCL_STATUS_SUCCESS;
+    CHIP_ERROR err       = CHIP_NO_ERROR;
+
+    chip::Platform::ScopedMemoryBuffer<uint8_t> cert;
+
+    uint8_t * certBuf = nullptr;
+
+    emberAfPrintln(EMBER_AF_PRINT_DEBUG, "OpCreds: commissioner has added an Op Cert");
+
+    AdminPairingInfo * admin = retrieveCurrentAdmin();
+    VerifyOrExit(admin != nullptr, status = EMBER_ZCL_STATUS_FAILURE);
+
+    VerifyOrExit(cert.Alloc(NOC.size() + ICACertificate.size()), status = EMBER_ZCL_STATUS_FAILURE);
+    certBuf = cert.Get();
+    memcpy(certBuf, NOC.data(), NOC.size());
+    memcpy(&certBuf[NOC.size()], ICACertificate.data(), ICACertificate.size());
+
+    VerifyOrExit(admin->SetOperationalCert(ByteSpan(certBuf, NOC.size() + ICACertificate.size())) == CHIP_NO_ERROR,
```

Comment:
We're allocating a temporary buffer in this function, then in SetOperationalCert allocating _another_ buffer just to make a copy.

We should either hand ownership of our existing buffer over (which is still wasteful if the admin already has a big enough buffer) or better yet have a version of SetOperationalCert that just takes the two separate bytespans for the NOC and ICA cert....

Reply:
This was done mainly because OCTET_STRING cannot carry data that are bigger than 256 bytes. And, op cert is not fitting into one. While Vivien is working on ZAP fix for that, to get this working, I had to split it into to arguments (i.e. hijack ICA, as it's not being used right now).

I had put a TODO on controller side to reflect this. I can probably copy/paste it here.

For cleaning up the ownership of the buffers, I would hope that we can do it in a follow up PR.

Reply:
Added a TODO for ownership of buffers.

Reply:
OK.  If this code is temporary, that's fine.